[mod] container: replace uWSGI with Granian

inetol opened the pull request:
The configuration in Granian is handled with ENVs, much more convenient and practical for updating. The settings have been tested for over two months in a production instance, being usable on small to somewhat large instances without having to modify anything.

It also removes the patch functions and ENVs abstraction from the entrypoint, this makes it possible to run the container with immutable configuration.

In some setups, It may be desired to have the volumes/files under a specific uid/gid (other than searxng:searxng), if the entrypoint has root permissions it will chown automatically on every start, which may not be desired. Explicitly setting the new ENV `FORCE_OWNERSHIP=false` will prevent ownership from being modified.

Those changes are reflected in the documentation, which also received a rework.

No manual migration is necessary **unless** the user has changed the default uWSGI configuration or has a very specific setup.

Closes https://github.com/searxng/searxng/issues/4894
Closes https://github.com/searxng/searxng/issues/4818
Closes https://github.com/searxng/searxng/issues/4802

Supersedes #4596

Related https://github.com/searxng/searxng/discussions/4479

---

Latest CI: https://github.com/inetol/searxng/actions/runs/15582123713

inetol committed:
[mod] container: replace uWSGI with Granian

The configuration in Granian is handled with ENVs, much more convenient and practical for updating. The settings have been tested for over two months in a production instance, being usable on small to somewhat large instances without having to modify anything.

It also removes the patch functions and ENVs abstraction from the entrypoint, this makes it possible to run the container with immutable configuration.

In some setups, It may be desired to have the volumes/files under a specific uid/gid (other than searxng:searxng), if the entrypoint has root permissions it will chown automatically on every start, which may not be desired. Explicitly setting the new ENV `FORCE_OWNERSHIP=false` will prevent ownership from being modified.

No manual migration is necessary **unless** the user has changed the default uWSGI configuration or has a very specific setup.

Closes https://github.com/searxng/searxng/issues/4894
Closes https://github.com/searxng/searxng/issues/4818
Closes https://github.com/searxng/searxng/issues/4802

Supersedes https://github.com/searxng/searxng/pull/4596

Related https://github.com/searxng/searxng/discussions/4479

inetol committed:
[mod] docs: add container/granian

All container documentation has been recreated.

A new documentation page has been created for Granian.

Bnyro commented on container/Dockerfile:
```diff
@@ -46,14 +42,23 @@ LABEL org.opencontainers.image.authors="searxng <$GIT_URL>" \
       org.opencontainers.image.version="$SEARXNG_GIT_VERSION"
 
 ENV SEARXNG_VERSION="$SEARXNG_GIT_VERSION" \
-    INSTANCE_NAME="SearXNG" \
-    AUTOCOMPLETE="" \
-    BASE_URL="" \
```
I assume that these variables should just be provided as runtime envs, e.g. `SEARXNG_BASE_URL`? I think we should also mention that in the `Environment variables` section of the newly added documentation, i.e. all env variables from https://github.com/searxng/searxng/blob/master/searx/settings_defaults.py#L179-L185.

I don't think we should copy all of them because we'd always need to keep the documentation in sync, but maybe we can reference that file somehow?

inetol replied:
ca70b86c075330598f4aaf31c6fcf8a4459a15c7

Bnyro commented on container/Dockerfile:
```diff
@@ -46,14 +42,23 @@ LABEL org.opencontainers.image.authors="searxng <$GIT_URL>" \
+    GRANIAN_RUNTIME_THREADS="3" \
+    GRANIAN_RUNTIME_BLOCKING_THREADS="2" \
+    GRANIAN_RESPAWN_INTERVAL="0" \
+    GRANIAN_WORKERS_KILL_TIMEOUT="30" \
```
```suggestion
    GRANIAN_WORKERS="1" \
    GRANIAN_WORKERS_KILL_TIMEOUT="30" \
```
I think it makes sense to add this for clarity, so that users can look through the envs here and can simply see how to set up more workers.

inetol replied:
I would prefer not to make workers option visible, as it induces increasing that value to solve the pains of life by following the bad practices we have now with uWSGI. If the server runs short, increase threads, not workers.

I said it in another issue, but increasing workers doubles the resource usage and reduces the effectiveness of botprotection (placebo?). The option is there documented, you can crank it up whenever you want, but I see this as the *not the best solution* for improving performance.

Bnyro replied:
> I would prefer not to make workers option visible, as it induces increasing that value to solve the pains of life by following the bad practices we have now with uWSGI. If the server runs short, increase threads, not workers.
>
In that case, please just mention that in the documentation as well, that's probably a good hint for people running large instances :)

inetol replied:
8ae618f9949787cf9b7ad8ea023a708bf5ef0a4c

Bnyro commented on container/Dockerfile:
```diff
@@ -46,14 +42,23 @@ LABEL org.opencontainers.image.authors="searxng <$GIT_URL>" \
+    GRANIAN_WEBSOCKETS="false" \
+    GRANIAN_LOOP="uvloop" \
+    GRANIAN_BLOCKING_THREADS="2" \
+    GRANIAN_RUNTIME_THREADS="3" \
```
I think this had previously set to `4` by default when we used uwsgi?

inetol replied:
I have used the smallest possible values without affecting server performance. Architecturally, Granian is *faster* than uWSGI, so this change makes sense.

Bnyro commented on container/Dockerfile:
```diff
@@ -46,14 +42,23 @@ LABEL org.opencontainers.image.authors="searxng <$GIT_URL>" \
+    GRANIAN_LOOP="uvloop" \
+    GRANIAN_BLOCKING_THREADS="2" \
+    GRANIAN_RUNTIME_THREADS="3" \
+    GRANIAN_RUNTIME_BLOCKING_THREADS="2" \
```
> blocking threads: the number of threads per worker interacting with the Python interpreter
>
> the default number of runtime threads and runtime blocking threads is fine for the vast majority of applications out there
>
The default seems to be 512 (https://github.com/emmett-framework/granian/blob/df0a44493488406b234838ae44dd8fc3ab1cc062/src/wsgi/serve.rs#L579). Can you explain why we set it to 2, please?

inetol replied:
I didn't see that it defaults to that many threads, I guess it's because of the synchronous nature of WSGI? I don't remember if there was another reason why I left this value so low, but I'll test this with 512 threads and see how it goes.

inetol replied:
Resolved in https://github.com/searxng/searxng/pull/4820#discussion_r2134827964

Bnyro commented on docs/admin/installation-granian.rst:
```diff
@@ -0,0 +1,50 @@
+
+Granian can be configured via option parameters and environment variables.
+
+We provide sane defaults that should fit all use cases,
```
```suggestion
We provide sane defaults that should fit most use cases,
```
I think that it makes sense to add that maintainers of large instances might need to adjust the settings a bit, I assume?

inetol replied:
3def38b22e0ff2b2614e30fb4c31412768767765

Bnyro reviewed (commented):
Looks awesome, can't wait to use it on my instance :)

I've left some small comments about the configuration, the general changes structure looks good to me though :)

inetol committed:
[enh] misc: apply suggestions

Minor documentation changes.

Suggested https://github.com/searxng/searxng/pull/4820#discussion_r2134539259
Suggested https://github.com/searxng/searxng/pull/4820#discussion_r2134538610
Suggested https://github.com/searxng/searxng/pull/4820#discussion_r2134827964
Suggested https://github.com/searxng/searxng/pull/4820#discussion_r2134544300
Suggested https://github.com/searxng/searxng/pull/4820#discussion_r2149387388

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

gi0baro commented on container/Dockerfile:
```diff
@@ -46,14 +42,23 @@ LABEL org.opencontainers.image.authors="searxng <$GIT_URL>" \
+    GRANIAN_RUNTIME_BLOCKING_THREADS="2" \
+    GRANIAN_RESPAWN_INTERVAL="0" \
+    GRANIAN_WORKERS_KILL_TIMEOUT="30" \
+    GRANIAN_BACKPRESSURE="1024" \
```
Granian maintainer here. After taking a look at the configuration, I'd rather suggest the following change:

```suggestion
    GRANIAN_BLOCKING_THREADS="4" \
    GRANIAN_WORKERS_KILL_TIMEOUT="30" \
```

Explanation:
- Given the app is WSGI, there's no need to change the Rust runtime threads configuration. Granian is fast enough to just use 1 runtime thread for serving http requests, as the bottleneck will always be the Python code in that regards.
- As you're using the static file serving, I won't change the number of the runtime blocking threads, as those threads will be used to read from disk and are managed in a dynamic pool by the runtime depending on the requests. Limiting those threads to such a low number will quite limit the throughput.
- With 1 worker, the respawn interval is noop, so you don't really need that. And, if you ever change the config to use more than 1 worker, I would still keep the default, as between spawning up the new one and killing the old one Granian would wait that amount of time. Having that delay to 0 means killing the previous worker even if the new one is not ready, and thus increasing the risk of dropping requests
- If you set the blocking threads, then I would leave the backpressure to Granian. Also, if you have 1 worker, that value is exactly the same of the backlog.

Also, keep in mind the number of blocking threads is as well affecting the maximum size of a dynamic pool, that will grow or shrink depending on the traffic. Given that, you might also want to tune the default `GRANIAN_BLOCKING_THREADS_IDLE_TIMEOUT` from 30s to a value that suits your needs more ;)

inetol replied:
Thanks for the feedback. Regarding `GRANIAN_BLOCKING_THREADS_IDLE_TIMEOUT`, I understand that there might be an overhead when spawning a new thread, but is there anything wrong with keeping it alive for a long time? (e.g. 300 sec)

gi0baro replied:
@inetol I'd say, since the load across the thread pool is usually quite balanced, such a high value basically mean Granian will ~never shrink the pool. Which can be a valid strategy, but really depends on the traffic shape. If you expect the traffic to stay at 0 for minutes and then spike up, it can be a good strategy. Otherwise, sounds a bit wasty to me :)

inetol replied:
a7e1d5dd92ddc1e7b56b5b280d20f370ffefa500

gi0baro commented on container/Dockerfile:
```diff
@@ -46,14 +42,20 @@ LABEL org.opencontainers.image.authors="searxng <$GIT_URL>" \
+    GRANIAN_HOST="::" \
+    GRANIAN_PORT="8080" \
+    # https://github.com/emmett-framework/granian/issues/568
+    GRANIAN_HTTP1_KEEP_ALIVE="false" \
```
*and just one more thing:* I would not disable keep alive, as it might improve latency especially in browsers. You already set the kill timeout, so you should be fine on shutdowns. You might even lower down that value to match what's the maximum response time you expect plus some margin (I'd say if your typical response time is in ms order, forcefully closing open connections after 5 seconds of inactivity is perfectly fine).

inetol replied:
I would prefer not to defer shutdown, what I don't know is whether setting `GRANIAN_WORKERS_KILL_TIMEOUT` would kill or gracefully stop the Python interpreter on timeout finish?

We don't care if we force close HTTP connections breaking active connections, but we don't want the interpreter to be SIGKILLed because other issues could arise.

gi0baro replied:
Understood.
fyi, 2.4 *might* improve keepalived connections shutdown.

inetol replied:
https://github.com/searxng/searxng/pull/4820/commits/cb15e70d20b42b3d9150f702798f34a476adfa7a

return42 reviewed (approved):
looks awesome :+1: .. I placed a WIP commit on top, we can squash before merge this PR .. In the WIP commit there are only minor changes to the documentation .. no significant changes.

Bnyro reviewed (approved):
I've been using this on my instance for the past week and didn't encounter any issues, so lgtm in that regard :+1:
